### PR TITLE
[SPARK-3162][MLlib] Local Tree Training Pt 1: Refactor RandomForest.scala into utility classes

### PR DESCRIPTION
**Pull request by smurching**
## What changes were proposed in this pull request?

Breaks up #19433 to help unblock #19666; after this PR is merged, #19666 can be merged.

This PR contains the changes made to migrate functionality from RandomForest.scala into the following utility classes:

* AggUpdateUtils
* ImpurityUtils
* SplitUtils

The PR also adds tests for split selection logic in TreeSplitUtilsSuite.

A follow-up PR will include the other changes from #19433:
* Local decision tree data structures & tests
* Local tree training logic & tests

## How was this patch tested?

Adds unit tests for split selection logic in TreeSplitUtilsSuite

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/tree/impl/TreeSplitUtilsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+  test("chooseUnorderedCategoricalSplit: basic case") {
+    val featureIndex = 0
+    // Construct data for unordered categorical feature
+    // label: 0 --> values: 1
+    // label: 1 --> values: 0, 2
+    // label: 2 --> values: 2
+    // Expected split: feature value 1 on the left, values (0, 2) on the right
+    val values = Array(1, 1, 0, 2, 2)
+    val featureArity = values.max + 1
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/tree/impl/TreeSplitUtilsSuite.scala]: In order to make the test more strict, can you increase the `featureArity`, `numExamples` and `numClasses` ? e.g., featureArity = 6 and numExamples = 10 and numClasses = 5

**Review by WeichenXu123 — COMMENTED** [review]
Good work. I mainly reviewed the new added testsuite part.

File: mllib/src/test/scala/org/apache/spark/ml/tree/impl/TreeSplitUtilsSuite.scala (hunk 1 above)
Comment (smurching) [reply mllib/src/test/scala/org/apache/spark/ml/tree/impl/TreeSplitUtilsSuite.scala]: @WeichenXu123 thanks for the feedback! Definitely agree that the test is a little weak right now.

IMO it's mainly weak due to the low feature arity (there only three possible splits, so the right one could be picked by chance). I think increasing the number of classes/examples substantially might make the test harder to reason about, but not opposed to that either - let me know what you think.

What about something like:

```
    val values = Array(0, 1, 2, 3, 2, 2, 4)
    val labels = Array(0.0, 0.0, 1.0, 1.0, 2.0, 2.0, 2.0)
    // label: 0 --> values: 0, 1
    // label: 1 --> values: 2, 3
    // label: 2 --> values: 2, 2, 4
    // Expected split: feature values (0, 1) on the left, values (2, 3, 4) on the right
```

This way we still test multiclass classification & test the split-selection logic more rigorously.

Comment (WeichenXu123) [reply mllib/src/test/scala/org/apache/spark/ml/tree/impl/TreeSplitUtilsSuite.scala]: I think it is OK. thanks!

**Comment by felixcheung** [comment]
ping?
I'm mostly interested in SPARK-3162

**Comment by wenbochang** [comment]
Any updates? this PR seems to address critical issue: https://issues.apache.org/jira/browse/SPARK-3162

**Pull request closed without merging**